src/core/runtime/dns-browser.js

```diff
@@ -1,33 +1,63 @@
-/* global self */
+/* eslint-env browser */
 'use strict'
 
+const TLRU = require('../../utils/tlru')
+const { default: PQueue } = require('p-queue')
+
+// Avoid sending multiple queries for the same hostname by caching results
+const cache = new TLRU(1000)
+// TODO: /api/v0/dns does not return TTL yet: https://github.com/ipfs/go-ipfs/issues/5884
+// However we know browsers themselves cache DNS records for at least 1 minute,
+// which acts a provisional default ttl: https://stackoverflow.com/a/36917902/11518426
+const ttl = 60 * 1000
+
+// browsers limit concurrent connections per host,
+// we don't want preload calls to exhaust the limit (~6)
+const _httpQueue = new PQueue({ concurrency: 4 })
+
+function unpackResponse (domain, response, callback) {
+  if (response.Path) {
+    return callback(null, response.Path)
+  } else {
+    const err = new Error(response.Message)
+    return callback(err)
+  }
+}
+
 module.exports = (domain, opts, callback) => {
   if (typeof opts === 'function') {
     callback = opts
     opts = {}
   }
-
   opts = opts || {}
-
   domain = encodeURIComponent(domain)
-  let url = `https://ipfs.io/api/v0/dns?arg=${domain}`
 
+  // `opts` impact returned value, so we cache per domain+opts
+  const query = `${domain}${JSON.stringify(opts)}`
+
+  // try cache first
+  if (!opts.nocache && cache.has(query)) {
+    const response = cache.get(query)
+    return unpackResponse(domain, response, callback)
+  }
+
+  // fallback to sending DNSLink query to ipfs.io
+  // TODO: replace this with generic DNS over HTTPS: https://github.com/ipfs/js-ipfs/issues/2212
+  let url = `https://ipfs.io/api/v0/dns?arg=${domain}`
   Object.keys(opts).forEach(prop => {
     url += `&${encodeURIComponent(prop)}=${encodeURIComponent(opts[prop])}`
   })
 
-  self.fetch(url, { mode: 'cors' })
+  _httpQueue.add(() => fetch(url, { mode: 'cors' })
```

Comment:
i would use https://github.com/sindresorhus/ky (check this discussion https://github.com/ipfs/js-ipfs-http-client/pull/1059#issuecomment-516378448) 

it will simplify the above lines too and you can specific retries

Reply:
Took a stab at `ky` in https://github.com/ipfs/js-ipfs/pull/2304/commits/410ac101935099f2fae4fa39fd0e5621cc2a079d. Code is _much_ simpler now. 

We were using `fetch` for preload and dns in browser, however  `addFromURL` was handled via custom code on top of `node-fetch`. I removed custom code around `fetch` and used `ky-universal` everywhere for consistency.

All of them benefit from retries provided by `ky` out of the box. 
Question: should we hardcode `timeout` and `retry`, or go with defaults?

Reply:
we will probably create an ky instance to be used everywhere but for now just leave it hardcoded